# CareerLens AI

## Notebook 02 – Technology Job Market Data Preprocessing

### 🎯 Objective

The objective of this notebook is to transform the raw Technology Job Market dataset into a structured technology-focused dataset suitable for analysis and AI-driven career recommendations.

The preprocessing workflow includes filtering technology-related careers, creating career domain classifications, validating the processed data, and exporting a reusable dataset for downstream analytical tasks.

### 📌 Background

The original dataset contains more than 1.6 million job postings across multiple industries.

Since CareerLens AI focuses exclusively on technology careers, a preprocessing stage is required to identify technology-related job postings and organize them into meaningful career domains.

The processed dataset generated in this notebook will serve as the foundation for all subsequent analyses and recommendation models.

## 📂 Preprocessing Workflow

1. Import Libraries
2. Load Raw Dataset
3. Technology Role Filtering
4. Validate Filtered Dataset
5. Create Career Domain Feature
6. Validate Career Domains
7. Export Processed Dataset
8. Summary & Next Steps

##🐍 Section 1 — Import Libraries

In [1]:
import pandas as pd

pd.set_option("display.max_columns", None)

##📂 Section 2 — Load Dataset

In [3]:
#Mount the google drive
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [4]:
#Load dataset
file_path = "/content/drive/MyDrive/CareerLensAI/job_descriptions.csv"

df = pd.read_csv(file_path)
df.head()

,Job Id,Experience,Qualifications,Salary Range,location,Country,latitude,longitude,Work Type,Company Size,Job Posting Date,Preference,Contact Person,Contact,Job Title,Role,Job Portal,Job Description,Benefits,skills,Responsibilities,Company,Company Profile
0,1089843540111562,5 to 15 Years,M.Tech,$59K-$99K,Douglas,Isle of Man,54.2361,-4.5481,Intern,26801,2022-04-24,Female,Brandon Cunningham,001-381-930-7517x737,Digital Marketing Specialist,Social Media Manager,Snagajob,Social Media Managers oversee an organizations...,"{'Flexible Spending Accounts (FSAs), Relocatio...","Social media platforms (e.g., Facebook, Twitte...","Manage and grow social media accounts, create ...",Icahn Enterprises,"{""Sector"":""Diversified"",""Industry"":""Diversifie..."
1,398454096642776,2 to 12 Years,BCA,$56K-$116K,Ashgabat,Turkmenistan,38.9697,59.5563,Intern,100340,2022-12-19,Female,Francisco Larsen,461-509-4216,Web Developer,Frontend Web Developer,Idealist,Frontend Web Developers design and implement u...,"{'Health Insurance, Retirement Plans, Paid Tim...","HTML, CSS, JavaScript Frontend frameworks (e.g...","Design and code user interfaces for websites, ...",PNC Financial Services Group,"{""Sector"":""Financial Services"",""Industry"":""Com..."
2,481640072963533,0 to 12 Years,PhD,$61K-$104K,Macao,"Macao SAR, China",22.1987,113.5439,Temporary,84525,2022-09-14,Male,Gary Gibson,9687619505,Operations Manager,Quality Control Manager,Jobs2Careers,Quality Control Managers establish and enforce...,"{'Legal Assistance, Bonuses and Incentive Prog...",Quality control processes and methodologies St...,Establish and enforce quality control standard...,United Services Automobile Assn.,"{""Sector"":""Insurance"",""Industry"":""Insurance: P..."
3,688192671473044,4 to 11 Years,PhD,$65K-$91K,Porto-Novo,Benin,9.3077,2.3158,Full-Time,129896,2023-02-25,Female,Joy Lucero,+1-820-643-5431x47576,Network Engineer,Wireless Network Engineer,FlexJobs,"Wireless Network Engineers design, implement, ...","{'Transportation Benefits, Professional Develo...",Wireless network design and architecture Wi-Fi...,"Design, configure, and optimize wireless netwo...",Hess,"{""Sector"":""Energy"",""Industry"":""Mining, Crude-O..."
4,117057806156508,1 to 12 Years,MBA,$64K-$87K,Santiago,Chile,-35.6751,-71.5429,Intern,53944,2022-10-11,Female,Julie Johnson,343.975.4702x9340,Event Manager,Conference Manager,Jobs2Careers,A Conference Manager coordinates and manages c...,"{'Flexible Spending Accounts (FSAs), Relocatio...",Event planning Conference logistics Budget man...,Specialize in conference and convention planni...,Cairn Energy,"{""Sector"":""Energy"",""Industry"":""Energy - Oil & ..."


##📋 Section 3 — Technology Role Filtering

In [45]:
it_roles = [
    "Software",
    "Developer",
    "Programmer",
    "Data Analyst",
    "Data Scientist",
    "Database",
    "Network",
    "Cyber",
    "Security",
    "Cloud",
    "DevOps",
    "QA",
    "Quality Assurance",
    "Machine Learning",
    "Artificial Intelligence",
    "Business Intelligence",
    "UX/UI",
    "Frontend",
    "Backend",
    "Business Analyst",
    "Systems Analyst",
    "IT Project Manager",
    "IT Manager",
    "Product Manager",
    "Agile",
    "Scrum"
]

tech_jobs = df[
    df["Role"].str.contains(
        "|".join(it_roles),
        case=False,
        na=False
    )
].copy()

##Validate Filtering

In [46]:
tech_jobs.shape

(275022, 23)

### Observation

Keyword-based filtering successfully extracted technology-related job postings from the original dataset.

Original Dataset:
1,615,940 records

Filtered Technology Dataset:
302,757 records

##Section 4 — Explore the filtered dataset

In [47]:
tech_jobs.head()

,Job Id,Experience,Qualifications,Salary Range,location,Country,latitude,longitude,Work Type,Company Size,Job Posting Date,Preference,Contact Person,Contact,Job Title,Role,Job Portal,Job Description,Benefits,skills,Responsibilities,Company,Company Profile
1,398454096642776,2 to 12 Years,BCA,$56K-$116K,Ashgabat,Turkmenistan,38.9697,59.5563,Intern,100340,2022-12-19,Female,Francisco Larsen,461-509-4216,Web Developer,Frontend Web Developer,Idealist,Frontend Web Developers design and implement u...,"{'Health Insurance, Retirement Plans, Paid Tim...","HTML, CSS, JavaScript Frontend frameworks (e.g...","Design and code user interfaces for websites, ...",PNC Financial Services Group,"{""Sector"":""Financial Services"",""Industry"":""Com..."
3,688192671473044,4 to 11 Years,PhD,$65K-$91K,Porto-Novo,Benin,9.3077,2.3158,Full-Time,129896,2023-02-25,Female,Joy Lucero,+1-820-643-5431x47576,Network Engineer,Wireless Network Engineer,FlexJobs,"Wireless Network Engineers design, implement, ...","{'Transportation Benefits, Professional Develo...",Wireless network design and architecture Wi-Fi...,"Design, configure, and optimize wireless netwo...",Hess,"{""Sector"":""Energy"",""Industry"":""Mining, Crude-O..."
5,116831420231957,4 to 12 Years,MCA,$59K-$93K,Brussels,Belgium,50.5039,4.4699,Full-Time,23196,2023-07-25,Male,Matthew Gill,(973)791-5355x52199,Software Tester,Quality Assurance Analyst,Snagajob,A Quality Assurance Analyst tests software and...,"{'Life and Disability Insurance, Stock Options...",Quality assurance processes Testing methodolog...,Test software applications and systems to iden...,Adani Ports and Special Economic Zone,"{""Sector"":""Infrastructure"",""Industry"":""Ports a..."
13,290761760113904,1 to 8 Years,B.Com,$56K-$86K,Banjul,Gambia,13.4432,-15.3101,Temporary,127900,2022-05-24,Female,Lisa Franklin,(229)625-8850x6916,Network Administrator,Network Security Analyst,FlexJobs,Protect an organizations computer networks and...,"{'Legal Assistance, Bonuses and Incentive Prog...",Network security protocols Intrusion detection...,"Manage and secure computer networks, including...",EnLink Midstream,"{""Sector"":""Energy"",""Industry"":""Pipelines"",""Cit..."
23,2808730603865187,2 to 14 Years,PhD,$62K-$124K,Tegucigalpa,Honduras,15.1990,-86.2419,Temporary,127385,2022-07-03,Both,Danny Nelson,3185179288,Software Engineer,Backend Developer,CareerBuilder,A Backend Developer is tasked with building an...,"{'Employee Assistance Programs (EAP), Tuition ...",Proficiency in one or more backend programming...,Develop server-side logic and databases using ...,Oil Search,"{""Sector"":""Energy/Oil and Gas"",""Industry"":""Oil..."


In [48]:
tech_jobs["Role"].value_counts().head(20)

,count
Role,
Network Administrator,17470
Quality Assurance Analyst,10541
Database Administrator,10482
Data Analyst,10406
Backend Developer,10404
Frontend Developer,10308
UX/UI Designer,7028
Network Security Analyst,7027
Data Scientist,7003


##Section 5 — Create Career Domains

### 🗂️ Career Domain Classification

The original dataset contains hundreds of technology-related job roles.
To enable meaningful career analysis, job roles were grouped into broader technology career domains.

This process is called **career taxonomy creation**, where related occupations are organized into structured categories based on their characteristics and required skills.

The Career Domain feature will support:
- Technology market analysis
- Skill demand comparison
- Career recommendations
- Personalized learning pathways

In [49]:
def classify_domain(role):

    role = role.lower()


    if any(keyword in role for keyword in [
        "backend",
        "frontend",
        "software",
        "developer",
        "programmer",
        "mobile",
        "web"
    ]):
        return "Software Development"


    elif any(keyword in role for keyword in [
        "data analyst",
        "data scientist",
        "business intelligence",
        "machine learning",
        "data architect"
    ]):
        return "Data & Analytics"


    elif any(keyword in role for keyword in [
        "cyber",
        "security",
        "penetration"
    ]):
        return "Cybersecurity"


    elif any(keyword in role for keyword in [
        "cloud",
        "devops"
    ]):
        return "Cloud & DevOps"


    elif any(keyword in role for keyword in [
        "network"
    ]):
        return "Networking"


    elif any(keyword in role for keyword in [
        "database",
        "sql"
    ]):
        return "Database"


    elif any(keyword in role for keyword in [
        "ux",
        "ui",
        "designer",
        "interaction"
    ]):
        return "UI/UX Design"


    elif any(keyword in role for keyword in [
        "business analyst",
        "system analyst",
        "business systems analyst",
        "systems analyst"
    ]):
        return "Business Analysis"

    elif any(keyword in role for keyword in [
    "scrum",
    "agile",
    "product owner",
    "release manager",
    "delivery manager",
    "agile coach"
    ]):
        return "Software Delivery & Agile Management"


    elif any(keyword in role for keyword in [
        "project manager",
        "it manager",
        "product manager",
        "technical product manager"
    ]):
        return "IT Management"


    elif any(keyword in role for keyword in [
        "qa",
        "quality assurance",
        "tester"
    ]):
        return "Quality Assurance"


    else:
        return "Other"

In [50]:
tech_jobs["Career Domain"] = tech_jobs["Role"].apply(classify_domain)

##Section 6 — Validate Career Domains

In [51]:
tech_jobs["Career Domain"].value_counts()

,count
Career Domain,
Software Development,90506
Networking,31454
Data & Analytics,31321
Cybersecurity,27914
Quality Assurance,24456
Database,17308
Cloud & DevOps,13886
Business Analysis,13779
IT Management,10410


In [52]:
tech_jobs.shape

(275022, 24)

In [34]:
tech_jobs.columns

Index(['Job Id', 'Experience', 'Qualifications', 'Salary Range', 'location',
       'Country', 'latitude', 'longitude', 'Work Type', 'Company Size',
       'Job Posting Date', 'Preference', 'Contact Person', 'Contact',
       'Job Title', 'Role', 'Job Portal', 'Job Description', 'Benefits',
       'skills', 'Responsibilities', 'Company', 'Company Profile',
       'Career Domain'],
      dtype='object')

In [35]:
tech_jobs.sample(10)

,Job Id,Experience,Qualifications,Salary Range,location,Country,latitude,longitude,Work Type,Company Size,Job Posting Date,Preference,Contact Person,Contact,Job Title,Role,Job Portal,Job Description,Benefits,skills,Responsibilities,Company,Company Profile,Career Domain
1534249,183007469032578,3 to 9 Years,B.Com,$58K-$109K,Andorra la Vella,Andorra,42.5063,1.5218,Contract,44567,2022-12-23,Both,Jonathan Goodwin,843-472-1323x452,Network Administrator,Network Engineer,Jobs2Careers,"Design, implement, and maintain computer netwo...","{'Transportation Benefits, Professional Develo...",Network infrastructure design Routing and swit...,"Design, configure, and maintain network infras...",National Grid,"{""Sector"":""Utilities"",""Industry"":""Utilities"",""...",Networking
1271094,712205071994014,0 to 14 Years,BCA,$63K-$102K,Kampala,Uganda,1.3733,32.2903,Full-Time,39421,2023-04-11,Both,Nicholas Nguyen,001-960-540-8625x572,Java Developer,Java Web Application Developer,USAJOBS,Java Web Application Developers create web app...,"{'Employee Referral Programs, Financial Counse...",Web application development Java web framework...,Focus on developing Java-based web application...,Marks & Spencer,"{""Sector"":""Retail"",""Industry"":""Retail - Genera...",Software Development
314028,2190715297403200,2 to 8 Years,BBA,$55K-$116K,Maseru,Lesotho,-29.6099,28.2336,Temporary,125556,2022-01-23,Male,Ian Long,271.732.1214,Software Engineer,Frontend Developer,Monster,A Frontend Developer is responsible for design...,"{'Transportation Benefits, Professional Develo...","Proficiency in HTML, CSS, and JavaScript Exper...",Design and implement user interfaces for web a...,Global Partners,"{""Sector"":""Energy"",""Industry"":""Wholesalers: Di...",Software Development
873932,167159868986069,5 to 15 Years,M.Com,$60K-$129K,Madrid,Spain,40.4637,-3.7492,Part-Time,77160,2023-02-19,Male,Tiffany Collins,+1-739-509-1126x296,Customer Service Manager,Service Quality Assurance Manager,Idealist,Service Quality Assurance Managers monitor and...,"{'Health Insurance, Retirement Plans, Paid Tim...",Quality assurance Quality control Process impr...,Ensure service quality and consistency by defi...,Intuit,"{""Sector"":""Financial Technology"",""Industry"":""C...",Quality Assurance
1162805,65202365403317,0 to 8 Years,BBA,$62K-$87K,N'Djamena,Chad,15.4542,18.7322,Full-Time,83822,2021-09-18,Both,Theresa Shaffer,627.495.5848,Database Administrator,Database Developer,Idealist,"A Database Developer designs, develops, and ma...","{'Tuition Reimbursement, Stock Options or Equi...",Database design and development SQL and databa...,"Design and implement database systems, tables,...","Lyft, Inc.","{""Sector"":""Transportation/Ridesharing"",""Indust...",Software Development
642063,1950687426563301,0 to 9 Years,M.Tech,$60K-$89K,Kuwait City,Kuwait,29.3117,47.4818,Full-Time,38123,2022-10-01,Both,Kelly Spears,613.400.4724x467,Data Analyst,Data Scientist,SimplyHired,Data Scientists use their expertise in data an...,"{'Employee Assistance Programs (EAP), Tuition ...",Machine learning algorithms and libraries (e.g...,Apply machine learning algorithms and statisti...,Starbucks,"{""Sector"":""Food and Beverage"",""Industry"":""Food...",Data & Analytics
429970,2046314174480520,3 to 15 Years,MCA,$61K-$95K,Victoria,Seychelles,-4.6796,55.4920,Part-Time,37850,2023-03-08,Male,Nicholas Martin,452-214-2566x51319,Database Administrator,Database Developer,Idealist,"A Database Developer designs, develops, and ma...","{'Life and Disability Insurance, Stock Options...",Database design and development SQL and databa...,"Design and implement database systems, tables,...",Celgene Corporation,"{""Sector"":""Pharmaceuticals"",""Industry"":""Pharma...",Software Development
717635,2147270983313107,5 to 13 Years,B.Tech,$65K-$83K,Copenhagen,Denmark,56.2639,9.5018,Intern,33688,2021-10-10,Both,Justin Tucker,001-814-351-5461,Web Developer,Frontend Web Developer,FlexJobs,Frontend Web Developers design and implement u...,"{'Child

In [53]:
tech_jobs["Career Domain"].value_counts()

,count
Career Domain,
Software Development,90506
Networking,31454
Data & Analytics,31321
Cybersecurity,27914
Quality Assurance,24456
Database,17308
Cloud & DevOps,13886
Business Analysis,13779
IT Management,10410


In [54]:
other_roles = tech_jobs[
    tech_jobs["Career Domain"] == "Other"
]

other_roles["Role"].value_counts().head(30)

,count
Role,


In [55]:
tech_jobs["Career Domain"].value_counts().sum()

np.int64(275022)

##Section 7 — Export Dataset

In [ ]:
#export version1
tech_jobs.to_csv( "/content/drive/MyDrive/CareerLensAI/technology_jobs.csv", index=False )

In [57]:
#export version2
tech_jobs.to_csv("/content/drive/MyDrive/CareerLensAI/technology_jobs_final.csv",index=False)

## 📌 Key Achievements

- Filtered technology-related job postings from the original dataset.
- Reduced the dataset from 1.6 million to 275,022 technology job postings.
- Created a structured Career Domain feature.
- Generated a reusable processed dataset for future analyses.

## 🚀 Next Steps

The processed dataset will be used to answer key business questions such as:

- Which technology careers have the highest demand?
- Which skills are most requested by employers?
- Which career domains are growing?
- How can CareerLens AI recommend personalized learning paths?